### ISSUE TITLE
When will Support for Windows 2016 Platform be officially available?

### ISSUE THREAD

**Issue by syazdi** [chef-cookbooks/windows#489]
### Cookbook version
[Version of the cookbook where you are encountering the issue]

### Chef-client version
[Version of chef-client in your environment]

### Platform Details
[Operating system distribution and release version. Cloud provider if running in the cloud]

### Scenario:
[What you are trying to achieve and you can't?]

### Steps to Reproduce:
[If you are filing an issue what are the things we need to do in order to repro your problem? How are you using this cookbook or any resources it includes?]

### Expected Result:
[What are you expecting to happen as the consequence of above reproduction steps?]

### Actual Result:
[What actually happens after the reproduction steps? Include the error output or a link to a gist if possible.]

**Comment by stuartpreston** [comment]
@syazdi Windows 2016 is supported by this cookbook, but many of the resources were written in the 2012-era and there may be a few resources that may not run correctly.  Can you describe what you're trying to achieve with the Windows cookbook that isn't working? 

If you have a general query it would be great to see you in the #windows channel on our [Chef Community Slack](http://community-slack.chef.io/)

**Comment by syazdi** [comment]
Hi @stuartpreston, thank you for your response.  So in my cookbook default.rb recipe file, I am trying to run a powershell_script to "Set Drive Labels" which runs just fine on a Windows 2012R2 server.  But when running on a Windows 2016 server, these are the errors I receive:


  ec2: * powershell_script[Set Drive Labels] action run
    ec2:
    ec2: ================================================================================
    ec2: Error executing action `run` on resource 'powershell_script[Set Drive Labels]'
    ec2: ================================================================================
    ec2:
    ec2: Mixlib::ShellOut::ShellCommandFailed
    ec2: ------------------------------------
    ec2: Expected process to exit with [0], but received '1'
    ec2: ---- Begin output of "C:\Windows\system32\WindowsPowerShell\v1.0\powershell.exe" -NoLogo -NonInteractive -NoProfile -ExecutionPolicy Bypass -InputFormat None -File "C:/Users/kapadmin/AppData/Local/Temp/chef-script20170808-3472-1fey227.ps1" ----
    ec2: STDOUT: Parameters not compatible.
    ec2: Parameters not compatible.
    ec2: Parameters not compatible.
    ec2: Parameters not compatible.
    ec2: STDERR:
    ec2: ---- End output of "C:\Windows\system32\WindowsPowerShell\v1.0\powershell.exe" -NoLogo -NonInteractive -NoProfile -ExecutionPolicy Bypass -InputFormat None -File "C:/Users/kapadmin/AppData/Local/Temp/chef-script20170808-3472-1fey227.ps1" ----
    ec2: Ran "C:\Windows\system32\WindowsPowerShell\v1.0\powershell.exe" -NoLogo -NonInteractive -NoProfile -ExecutionPolicy Bypass -InputFormat None -File "C:/Users/kapadmin/AppData/Local/Temp/chef-script20170808-3472-1fey227.ps1" returned 1
    ec2:
    ec2: Resource Declaration:
    ec2: ---------------------
    ec2: # In c:/windows/temp/packer-chef-solo/local-mode-cache/cache/cookbooks/ktp-base-sql/recipes/default.rb
    ec2:
    ec2: 7: powershell_script 'Set Drive Labels' do
    ec2: 8:   code <<-EOH
    ec2: 9:   label C: System
    ec2: 10:   label D: SQLFILES
    ec2: 11:   label E: SQLDATA
    ec2: 12:   label F: SQLLOGS
    ec2: 13:   label G: SQLTEMPDB
    ec2: 14:   EOH
    ec2: 15: end
    ec2: 16:
    ec2:
    ec2: Compiled Resource:
    ec2: ------------------
    ec2: # Declared in c:/windows/temp/packer-chef-solo/local-mode-cache/cache/cookbooks/ktp-base-sql/recipes/default.rb:7:in `from_file'

**Comment by mwrock** [comment]
Looks like the actual `powershell_script` is executing but the `label` command is returning `Parameters not compatible` and an unsuccesful exit code. Its possible that those drives are just not available. Have you tried issuing those same commands on the node outside of chef?

**Comment by syazdi** [comment]
Hi @mwrock I ran those commands on the node outside of chef and they work just fine. If you look at the above errors more closely, the error is in line 7 which is the "powershell_script" command itself:

"ec2: 7: powershell_script 'Set Drive Labels' do"

ec2: * powershell_script[Set Drive Labels] action run
"ec2: Error executing action run on resource 'powershell_script[Set Drive Labels]'

So not sure if this is an issue with Windows 2016, because once again this script runs just fine on Windows 2012R2.  And as I said these commands were successful when run manually on a Windows 2016 server outside of chef.  So is this just some issue of Chef Windows cookbook compatibility with Windows 2016 servers?

**Comment by stuartpreston** [comment]
The Windows cookbook is not involved here at all, @syazdi as powershell_script is built into Core Chef. 

I tested this on my Windows 2016 Server and was only able to get the behavior seen when running label against a non existant drive (e.g Z:). I am wondering how the chef client is run on your system (manually, scheduled task or service?) and what user it it runs under? I will try and replicate your set up.

**Comment by syazdi** [comment]
@stuartpreston -  I'm using Packer to build a Windows Server 2016 ami from ec2 builder. Here is the block of code where we get packer to build the image using the cookbook:

      else if [[ "${GUEST_OS_TYPE}" =~ windows ]] && [[ "${COOKBOOK}" =~ ktp-base-sql ]] ; then
        packer validate ${BUILDSTRING} -var guest_os_type=${GUEST_OS_TYPE} -var cookbook=${COOKBOOK} -var commit=${GIT_COMMIT_HASH} ./buildpacks/build_sql.json || { message "Error validating buildpack. Aborted."; exit 1; }
        message "Done!"
        message "Building..."
        packer build ${BUILDSTRING} -var guest_os_type=${GUEST_OS_TYPE} -var cookbook=${COOKBOOK} -var commit=${GIT_COMMIT_HASH} ./buildpacks/build_sql.json || { message "Error building ${COOKBOOK}. Aborted."; exit 1; }


And to answer your second question, it looks like we are using Chef solo:

==> ec2: Connected to WinRM!
==> ec2: Provisioning with chef-solo
    ec2: Installing Chef...
    ec2:
    ec2: ModuleType Version    Name                                ExportedCommands
    ec2: ---------- -------    ----                                ----------------
    ec2: Script     0.0        Omnitruck                           {Get-ProjectMetadata, Install-Project, install}
    ec2: Installing chef from C:\Users\kapadmin\AppData\Local\Temp\chef-client-13.2.20-1-x64.msi
    ec2:
    ec2:
    ec2: Creating directory: C:/Windows/Temp/packer-chef-solo

**Comment by stuartpreston** [comment]
According to the error message and this script, the cookbook in use containing the `powershell_script` block is `ktp-base-sql` not the `windows` cookbook - could you verify?

**Comment by syazdi** [comment]
@stuartpreston - The windows cookbook is added in as a packaged cookbook, and in fact we have to use " cookbook 'windows', '= 2.1.1' " as we were facing issues with "cookbook 'windows', '~> 3.1.1'"  particularly this one:

 ec2: NoMethodError
    ec2: -------------
    ec2: undefined method `success_codes' for Chef::Resource::WindowsPackage

Once we go back to " cookbook 'windows', '= 2.1.1' ", the error is gone, but I can always open a different ISSUE to troubleshoot that.  Don't want to lose focus on what we are discussing here. But yes we are definitely using the Windows cookbook, that is built in into all our Windows related Chef cookbooks.

**Comment by stuartpreston** [comment]
@syazdi This is the repo for the Windows cookbook.  You've shown us an error that seems to occur on Windows 2016 when using the `powershell_script` resource.  That resource is part of core Chef and not part of the Windows cookbook.  Additionally as I mentioned, I have successfully used `powershell_script` to execute label commands on Windows 2016 and it works correctly.  Having a dependency between a cookbook and a specific version of the Windows cookbook is fairly normal.  

The error you see is originating on line 7 of your `ktp-base-sql` cookbook's default recipe (default.rb).  This may well be pulled in as a dependency from another cookbook (but we cannot see the detail of that from what you have supplied in the issue so far).

# In c:/windows/temp/packer-chef-solo/local-mode-cache/cache/cookbooks/**ktp-base-sql/recipes/default.rb**

```
ec2:
ec2: 7: powershell_script 'Set Drive Labels' do
ec2: 8: code <<-EOH
ec2: 9: label C: System
ec2: 10: label D: SQLFILES
ec2: 11: label E: SQLDATA
ec2: 12: label F: SQLLOGS
ec2: 13: label G: SQLTEMPDB
ec2: 14: EOH
ec2: 15: end
ec2: 16:
```

TL;DR - you should be able to replicate your specific issue without using the Windows cookbook at all.  If you see one of the resources from Windows cookbook failing on Windows 2016, please provide more details so we can investigate further.   Thanks!

**Comment by syazdi** [comment]
Hi @stuartpreston I saw this solution offered on one of the other ISSUE topics:

The packer chef solo provisioner simply runs chef client via a winrm remote session. Certain tasks such as installing some windows features or updates will always fail in a remote session. The common way to work around that is to use a scheduled task via the remote session. Your remote session creates and runs the scheduled task and the task executes in a local security context. The chef solo provisioner does not employ that work around. However the powershell provisioner does as long as you provide credentials.

Do you think using "powershell provisioner with elevated rights" might be a potential solution? Although I can understand from your above explanation that you are not seeing the correlation, but let's just assume there is a correlation, do you recommend that as a solution?

**Comment by stuartpreston** [comment]
@syazdi Yes that's quite possible, there's a lot of things that behave irrationally when using WinRM to install them remotely.  As you mention, we work around this in things like Test Kitchen by providing our own provisioner that can run elevated from a Scheduled Task.  If the packer powershell provisioner already provides this mechanism of transferring files to the remote machine and then executing a command running elevated, then that could be a good way of getting closer to a system that behaves as good as running Chef Client locally on the machine.  Especially if you plan to do this a lot in future.

**Comment by stuartpreston** [comment]
@syazdi I'm going to close this issue for now as there doesn't appear to be anything actionable. Please open a new issue with details if you have a specific issue with the Windows cookbook on Windows Server 2016.
